jmespath-core/src/main/java/io/burt/jmespath/node/JsonLiteralNode.java

```diff
@@ -3,35 +3,33 @@
 import io.burt.jmespath.Adapter;
 
 public class JsonLiteralNode<T> extends Node<T> {
-  private final String raw;
-  private final T tree;
+  private final String rawJson;
+  private final T json;
 
-  public JsonLiteralNode(Adapter<T> runtime, String raw, T tree) {
+  public JsonLiteralNode(Adapter<T> runtime, String rawJson) {
     super(runtime);
-    this.raw = raw;
-    this.tree = tree;
+    this.rawJson = rawJson;
+    this.json = runtime.parseString(rawJson);
   }
 
   @Override
   public T search(T input) {
-    return tree;
+    return json;
   }
 
   @Override
   protected String internalToString() {
-    return String.format("%s", raw);
+    return String.format("%s", rawJson);
   }
 
   @Override
   protected boolean internalEquals(Object o) {
     JsonLiteralNode<?> other = (JsonLiteralNode<?>) o;
-    return tree.equals(other.tree);
+    return json.equals(other.json);
```

Comment:
I debated back and forth about this yesterday. This is the way it's been before so in the end I opted not to change. There are pros and cons to both alternatives.

I think it makes sense to compare the parsed JSON value rather than the string since we're mostly interested in equivalence, if two nodes are exchangeable. `#equals` is also exclusively used by the tests, so far, and I think it only makes sense if two nodes are still equal if you happen to put spaces in the assertion but not in the expression, for example.

The downside is that we assume that the JSON representation we're working with has a proper implementation of `#equals` (which isn't necessarily true, I'm not entirely sure about our own JSON parser in the JCF runtime). Perhaps we should do `runtime.compare(json, other.json) == 0` to be sure.

`StringNode#equals` does the opposite, it compares the raw value and not the parsed value. The reason is that I don't see a scenario in which the raw and parsed value are not equal, and using `String#equals` feels safer than relying on `T#equals` being correct (but again, perhaps `runtime.compare(string, other.string) == 0` is the right thing to do, it just feels a bit overkill).